Mathlib/Tactic/Simp/FinsetInterval.lean

```diff
@@ -0,0 +1,183 @@
+/-
+Copyright (c) 2025 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import Mathlib.Order.Interval.Finset.Nat
+import Mathlib.Util.Qq
+import Mathlib.Tactic.NormNum.Basic
+
+/-!
+# Simproc for intervals of natural numbers
+-/
+
+open Qq Lean Mathlib.Meta.NormNum Finset
+
+namespace Mathlib.Tactic.Simp
+section lemmas
+variable {m n : ℕ} {s : Finset ℕ}
+
+private lemma Icc_eq_insert_of_Icc_succ_eq (hmn : m ≤ n) (hs : Icc (m + 1) n = s) :
+    Icc m n = insert m s := by rw [← hs, Nat.Icc_insert_succ_left hmn]
+
+private lemma Ico_eq_of_Icc_pred_eq (hn : n ≠ 0) (hs : Icc m (n - 1) = s) : Ico m n = s := by
+  rw [← hs, Nat.Icc_pred_right _ hn.bot_lt]
+
+private lemma Ico_zero (m : ℕ) : Ico m 0 = ∅ := by simp
+
+private lemma Ioc_eq_of_Icc_succ_eq (hs : Icc (m + 1) n = s) : Ioc m n = s := by
+  rw [← hs, Nat.Icc_succ_left]
+
+private lemma Ioo_eq_of_Icc_succ_pred_eq (hs : Icc (m + 1) (n - 1) = s) : Ioo m n = s := by
+  rw [← hs, ← Nat.Icc_succ_pred_eq_Ioo]
+
+private lemma Iic_eq_of_Icc_zero_eq (hs : Icc 0 n = s) : Iic n = s := hs
+
+private lemma Iio_eq_of_Icc_zero_pred_eq (hn : n ≠ 0) (hs : Icc 0 (n - 1) = s) : Iio n = s :=
+  Ico_eq_of_Icc_pred_eq hn hs
+
+private lemma Iio_zero : Iio 0 = ∅ := by simp
+
+end lemmas
+
+/-- Given natural numbers `m` and `n`, returns `(s, ⊢ Finset.Icc m n = s)`. -/
+partial def evalFinsetIccNat {em eml en enl : Q(ℕ)} (m n : ℕ) (hm : Q(IsNat $em $eml))
+    (hn : Q(IsNat $en $enl)) : MetaM ((s : Q(Finset ℕ)) × Q(.Icc $em $en = $s)) := do
+  -- If `m = n`, then `Icc m n = {m}`. We handle this case separately because `insert m ∅` is
+  -- not synteq to `{m}`.
+  if m = n then
+    have : $em =Q $en := ⟨⟩
+    return ⟨q({$em}), q(Icc_self _)⟩
+  -- If `m < n`, then `Icc m n = insert m (Icc m n)`.
+  else if m < n then
+    let hmn ← mkDecideProofQq q($em ≤ $en)
+    let hm := q(isNat_natSucc $hm rfl)
+    let ⟨s, hs⟩ ← evalFinsetIccNat (m+1) n hm hn
+    return ⟨q(insert $em $s), q(Icc_eq_insert_of_Icc_succ_eq $hmn $hs)⟩
+  -- Else `n < m` and `Icc m n = ∅`.
+  else
+    let hnm ← mkDecideProofQq q($en < $em)
+    return ⟨q(∅), q(Finset.Icc_eq_empty_of_lt $hnm)⟩
+
+end Mathlib.Tactic.Simp
+
+open Mathlib.Tactic.Simp
+
+/-!
+Note that these simprocs are not made simp to avoid simp blowing up on goals containing things of
+the form `Iic (2 ^ 1024)`.
+-/
```

Comment:
Something to consider: turn these on by default, but with a hard-coded (or option controlled?) limit on the size of the produced Finset. (Perhaps around 37? :-)

There's a trade-off between convenience, and someone getting confused about why it suddenly stops working when they use it above the limit.

Reply:
I don't think it is ever convenient for these to be on by default? Happy to be outvoted here